src/gui/dragAndDrop.js

```diff
@@ -2,48 +2,197 @@
+
+    const backendNode = NodeForest.getNode(node.id);
+    // Insert into a binary node
+    if (target.id.toString().endsWith("Left")) {
+        NodeForest.getNode(target.parentNode.id).setLeftOperand(backendNode);
+    }
+    else if (target.id.toString().endsWith("Right")) {
+        NodeForest.getNode(target.parentNode.id).setRightOperand(backendNode);
+    }
+    target.appendChild(node);
+}
+
+function handleChildNodeEquationPane(node, target, event) {
+    const backendNode = NodeForest.getNode(node.id);
+
+    // Remove from binary operator
+    if (target.id.toString().endsWith("Left")) {
+        NodeForest.getNode(node.parentNode.parentNode.id).setLeftOperand(undefined);
+    }
+    else if (target.id.toString().endsWith("Right")) {
+        NodeForest.getNode(node.parentNode.parentNode.id).setLeftOperand(undefined);
+    }
+    node.parentNode.removeChild(node);
+
+    NodeForest.insertRootNode(backendNode);
+
+    target.appendChild(node);
+    node.style.position = "fixed";
+    node.style.top = `${event.pageY}px`;
+    node.style.left = `${event.pageX}px`;
+}
+
+/**
+ * Determines how to handle the dropping of each Node that is dragged into the editor space.
+ * @param event
+ */
 function dropNode(event) {
     event.preventDefault();
+    event.stopPropagation();
     const id = event.dataTransfer.getData("text");
-    const parentId = event.dataTransfer.getData("parent");
     const node = document.getElementById(id);
+    const target = event.target;
+    let origin;
+    let destination;
+    const handlers = {
+        panel: {
+            node: handlePanelNode,
+            equationPane: handlePanelEquationPane
+        },
+        rootNode: {
+            node: handleRootNodeNode,
+            equationPane: handleRootNodeEquationPane
+        },
+        childNode: {
+            node: handleChildNodeNode,
+            equationPane: handleChildNodeEquationPane
+        }
+    };
+
+    // determine origin
+    if (node.parentNode.classList.contains("panel")) {
+        origin = "panel";
+    }
+    else if (node.parentNode.classList.contains("node-container")) {
+        origin = "childNode";
+    }
+    else {
+        origin = "rootNode";
+    }
 
-    // Determine behavior based upon source dragged from
-    if (parentId !== "editorPane") {
-        const clone = node.cloneNode(true);
-        clone.id = NodeFactory.createNode(clone.classList, true);
-        clone.style.position = "fixed";
-        clone.style.top = `${event.pageY}px`;
-        clone.style.left = `${event.pageX}px`;
-        event.target.appendChild(clone);
+    // determine destination
+    if (target.classList.contains("node-container")) {
+        destination = "node";
     }
     else {
-        node.style.position = "fixed";
-        node.style.top = `${event.pageY}px`;
-        node.style.left = `${event.pageX}px`;
+        destination = "equationPane";
     }
 
+    handlers[origin][destination](node, target, event);
```

Comment:
What is going on here?

Reply:
It maps to the appropriate handler and then calls it. handlers contains a hashmap of <String><String><Functions> basically. The first one is the origin of the node (is it coming from the left pane, another node, etc.) The second one is the destination. Then it calls the appropriate function using this and passes in the node target and event so it can handle it.